group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -900,9 +899,8 @@ private void append(
                 if (currentBatch != null) {
                     currentBatch.deferredEvents.add(event);
                 } else {
-                    OptionalLong pendingOffset = deferredEventQueue.highestPendingOffset();
-                    if (pendingOffset.isPresent()) {
-                        deferredEventQueue.add(pendingOffset.getAsLong(), event);
+                    if (coordinator.lastCommittedOffset() < coordinator.lastWrittenOffset()) {
```

Comment:
Pardon me, does it means "there is no pending offset if `lastCommittedOffset` == `lastWrittenOffset``?

Reply:
Correct. The last written offset is basically the log end offset of the partition at the time of the last write and the last committed offset is the high watermark. If they are equal, it means that there are not pending writes (or uncommitted records).